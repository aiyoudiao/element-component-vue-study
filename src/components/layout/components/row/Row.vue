<script setup lang="ts">
import { computed, getCurrentInstance, provide, toRefs, unref, withDefaults } from 'vue'

interface RowType {
  tag?: string
  gutter?: number
  type?: string
  justify?: string
  align?: string
}

interface StyleType {
  marginLeft?: string
  marginRight?: string
}

const props = withDefaults(defineProps<RowType>(), {
  tag: 'div',
  gutter: 0,
  justify: 'start',
  align: 'top',
})

const { tag } = toRefs<RowType>(props)

function useClassesAndStyle(props: RowType) {
  const classList = ['el-row']
  const style = computed<StyleType>(() => {
    const result: StyleType = {}
    const gutter = unref<number>(props.gutter)
    if (gutter) {
      result.marginLeft = `-${gutter / 2}px`
      result.marginRight = result.marginLeft
    }
    return result
  })

  const { justify, align, type: typeName } = props

  if (unref(justify) !== 'start')
    classList.push(`is-justify-${unref(justify)}`)

  if (unref(align) !== 'top')
    classList.push(`is-align-${unref(align)}`)

  if (unref(typeName) === 'flex')
    classList.push('el-row--flex')

  provide('el-row', getCurrentInstance())
  return [classList.join(' '), unref(style)]
}

const [classes, style] = useClassesAndStyle(props)
</script>

<script lang="ts">
export default {
  name: 'ElRow',
}
</script>

<template>
  <component :is="unref(tag)" :class="classes" :style="style">
    <slot />
  </component>
</template>

<style lang="scss">
@import '../../../../theme/src/row.scss';
</style>

