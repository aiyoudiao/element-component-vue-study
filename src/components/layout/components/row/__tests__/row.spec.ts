import { mount } from '@vue/test-utils'
import Row from '../Row.vue'

describe('Row.vue', () => {
  it('Row snapshot', () => {
    const wrapper = mount(Row)
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('Row class', () => {
    it('contain has class .el-row', () => {
      const wrapper = mount(Row)
      expect(wrapper.find('.el-row').exists()).toBe(true)
    })
    it('contain has class .el-row--flex', () => {
      const wrapper = mount(Row, {
        props: {
          type: 'flex',
        },
      })
      expect(wrapper.find('.el-row--flex').exists()).toBe(true)
    })
    it('contain has class .is-justify-end', () => {
      const wrapper = mount(Row, {
        props: {
          justify: 'end',
        },
      })
      expect(wrapper.find('.is-justify-end').exists()).toBe(true)
    })
    it('contain has class .is-align-bottom', () => {
      const wrapper = mount(Row, {
        props: {
          align: 'bottom',
        },
      })
      expect(wrapper.find('.is-align-bottom').exists()).toBe(true)
    })
  })

  it('Row gutter style', () => {
    const count = 10
    const gutter = count * 2
    const wrapper = mount(Row, {
      props: {
        gutter,
      },
    })
    expect(wrapper.attributes('style')).toBe(`margin-left: -${count}px; margin-right: -${count}px;`)
  })
})
