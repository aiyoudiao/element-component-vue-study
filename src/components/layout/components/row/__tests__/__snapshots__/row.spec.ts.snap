// Vitest Snapshot v1

exports[`Row.vue > Row snapshot 1`] = `
<div
  class="el-row"
>
  
  
</div>
`;
